src/OpenTelemetry.Instrumentation.AspNet.Win/Implementation/HttpInListener.cs

```diff
@@ -154,15 +172,36 @@ public override void OnStopActivity(Activity activity, object payload)
                 else if (routeData.Route is Route route)
                 {
                     // MVC + WebAPI traditional routing & MVC attribute routing flow here.
-
                     template = route.Url;
                 }
 
                 if (!string.IsNullOrEmpty(template))
                 {
                     // Override the name that was previously set to the path part of URL.
-                    activity.DisplayName = template;
-                    activity.AddTag(SpanAttributeConstants.HttpRouteKey, template);
+                    activityToEnrich.DisplayName = template;
+                    activityToEnrich.AddTag(SpanAttributeConstants.HttpRouteKey, template);
+                }
+            }
+
+            if (!(this.options.TextFormat is TraceContextFormatActivity))
+            {
+                if (activity.OperationName.Equals(ActivityNameByHttpInListener))
+                {
+                    // If instrumentation started a new Activity, it must
+                    // be stopped here.
+                    activity.Stop();
+
+                    // Restore the original activity as Current.
+                    var activityByAspNet = (Activity)activity.GetCustomProperty("ActivityByAspNet");
+                    Activity.Current = activityByAspNet;
```

Comment:
When we call activity.Stop() on the one we created it will do Activity.SetCurrent(Parent) (which should be the libraryActivity) so I don't think we need to do this Activity.Current set?

Reply:
We need it because the newactivity is created *not* as the child of the libraryActivity, but rather as a "sibling" of it. i.e having same parent as libraryActivity.
`newOne.SetParentId(ctx.TraceId, ctx.SpanId, ctx.TraceFlags);`

Reply:
Ah makes sense thanks.